File: WPILibShortcutCreator/src/main/cpp/ShortcutCreator.cpp
#include "ShortcutCreator.h"

#include <optional>


#include "shobjidl.h"
#include "objidl.h"
#include "shlguid.h"
#include "Shlobj.h"
#include "Shlobj_core.h"
#include "ComPtr.h"

#include "CreateDirectoryRecursively.h"

#pragma comment(lib, "Shell32.lib")

ShortcutCreator::ShortcutCreator() {
  m_hr = CoGetClassObject(CLSID_ShellLink, CLSCTX_INPROC_SERVER, NULL, IID_PPV_ARGS(shellLinkFactory.GetAddressOf()));
}

bool ShortcutCreator::CreateShortcut(const std::wstring& destination, const ShortcutInfo& shortcutInfo) const {
  HRESULT hres;
  com::ComPtr<IShellLink> shellLink;

  hres = shellLinkFactory->CreateInstance(NULL, IID_PPV_ARGS(shellLink.GetAddressOf()));

  if (FAILED(hres)) return false;

  auto persistFile = shellLink.As<IPersistFile>();

  if (!persistFile) return false;

  shellLink->SetPath(shortcutInfo.path.c_str());

  shellLink->SetDescription(shortcutInfo.description.c_str());

  if (!shortcutInfo.iconLocation.empty()) {
    shellLink->SetIconLocation(shortcutInfo.iconLocation.c_str(), 0);
  }

  std::wstring finalPath = destination + L'\\' + shortcutInfo.name + L".lnk";

  hres = persistFile->Save(finalPath.c_str(), true);

  if (hres == HRESULT_FROM_WIN32(ERROR_PATH_NOT_FOUND)) {
    static const std::wstring separators(L"\\/");
    std::size_t slashIndex = finalPath.find_last_of(separators);
    if(slashIndex != std::wstring::npos) {
      bool wasSuccessful = createDirectoryRecursively(finalPath.substr(0, slashIndex));
      if (wasSuccessful) {
        hres = persistFile->Save(finalPath.c_str(), true);
      } else {
        return false;
      }
    } else {
      return false;
    }
  }

  return SUCCEEDED(hres);
}

std::optional<std::wstring> ShortcutCreator::GetLocalStartMenuFolder() const {
  PWSTR path;
  if (SUCCEEDED(SHGetKnownFolderPath(FOLDERID_StartMenu, 0, NULL, &path))) {
    std::wstring retVal{path};
    CoTaskMemFree(path);
    return retVal;
  }
  return nullptr;
}

std::optional<std::wstring> ShortcutCreator::GetAllUsersStartMenuFolder() const {
  PWSTR path;
  if (SUCCEEDED(SHGetKnownFolderPath(FOLDERID_CommonStartMenu, 0, NULL, &path))) {
    std::wstring retVal{path};
    CoTaskMemFree(path);
    return retVal;
  }
  return nullptr;
}

std::optional<std::wstring> ShortcutCreator::GetPublicDesktopFolder() const {
  PWSTR path;
  if (SUCCEEDED(SHGetKnownFolderPath(FOLDERID_PublicDesktop, 0, NULL, &path))) {
    std::wstring retVal{path};
    CoTaskMemFree(path);
    return retVal;
  }
  return nullptr;
}

std::optional<std::wstring> ShortcutCreator::GetLocalDesktopFolder() const {
  PWSTR path;
  if (SUCCEEDED(SHGetKnownFolderPath(FOLDERID_Desktop, 0, NULL, &path))) {
    std::wstring retVal{path};
    CoTaskMemFree(path);
    return retVal;
  }
  return nullptr;
}

bool ShortcutCreator::CreateFolder(const std::wstring& path) const {
  auto ret = CreateDirectory(path.c_str(), NULL);
  return ret == 0;
}

bool ShortcutCreator::CreateShortcuts(std::vector<ShortcutInfo>& toCreate, const std::wstring& dest) const {
  bool allCompleted = true;

  for (auto&& info : toCreate) {
    if (!CreateShortcut(dest, info)) {
      allCompleted = false;
    }
  }

  return allCompleted;
}

// bool ShortcutCreator::Create() const {
//   HRESULT hres;
//   com::ComPtr<IShellLink> shellLink;

//   hres = CoCreateInstance(CLSID_ShellLink, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(shellLink.GetAddressOf()));

//   if (FAILED(hres)) return false;


//   auto persistFile = shellLink.As<IPersistFile>();
// }

File: WPILibShortcutCreator/ShortcutCreator.cs
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.Marshalling;
using WPILibInstaller.Models;

namespace WPILibShortcutCreator;

internal static class ShortcutCreator
{
    private static readonly StrategyBasedComWrappers ComWrappers = new();

    public static bool CanCreateShellLinks()
    {
        try
        {
            object? shellLink = CreateShellLinkObject();
            return shellLink is IShellLinkW and IPersistFile;
        }
        catch (COMException)
        {
            return false;
        }
    }

    public static bool CreateShortcuts(IReadOnlyList<ShortcutInfo> shortcuts, string destination)
    {
        bool allCompleted = true;

        foreach (ShortcutInfo shortcut in shortcuts)
        {
            if (!CreateShortcut(destination, shortcut))
            {
                allCompleted = false;
            }
        }

        return allCompleted;
    }

    private static bool CreateShortcut(string destination, ShortcutInfo shortcutInfo)
    {
        try
        {
            object? shellLinkObject = CreateShellLinkObject();
            if (shellLinkObject is not IShellLinkW shellLink
                || shellLinkObject is not IPersistFile persistFile)
            {
                return false;
            }

            if (HResult.Failed(shellLink.SetPath(shortcutInfo.Path))
                || HResult.Failed(shellLink.SetDescription(shortcutInfo.Description)))
            {
                return false;
            }

            if (!string.IsNullOrEmpty(shortcutInfo.IconLocation)
                && HResult.Failed(shellLink.SetIconLocation(shortcutInfo.IconLocation, 0)))
            {
                return false;
            }

            string finalPath = Path.Combine(destination, shortcutInfo.Name + ".lnk");
            int hresult = persistFile.Save(finalPath, remember: true);

            if (hresult == HResult.FromWin32(NativeMethods.ErrorPathNotFound)
                && TryCreateParentDirectory(finalPath))
            {
                hresult = persistFile.Save(finalPath, remember: true);
            }

            return HResult.Succeeded(hresult);
        }
        catch (COMException)
        {
            return false;
        }
        catch (InvalidCastException)
        {
            return false;
        }
    }

    private static object? CreateShellLinkObject()
    {
        Guid classId = ComGuids.ShellLinkClassId;
        Guid interfaceId = typeof(IShellLinkW).GUID;
        int hresult = NativeMethods.CoCreateInstance(
            ref classId,
            IntPtr.Zero,
            NativeMethods.ClsctxInprocServer,
            ref interfaceId,
            out IntPtr shellLinkPointer);

        if (HResult.Failed(hresult))
        {
            return null;
        }

        try
        {
            return ComWrappers.GetOrCreateObjectForComInstance(
                shellLinkPointer,
                CreateObjectFlags.UniqueInstance);
        }
        finally
        {
            _ = Marshal.Release(shellLinkPointer);
        }
    }

    private static bool TryCreateParentDirectory(string filePath)
    {
        try
        {
            string? parentDirectory = Path.GetDirectoryName(filePath);
            if (string.IsNullOrEmpty(parentDirectory))
            {
                return false;
            }

            Directory.CreateDirectory(parentDirectory);
            return true;
        }
        catch (ArgumentException)
        {
            return false;
        }
        catch (IOException)
        {
            return false;
        }
        catch (NotSupportedException)
        {
            return false;
        }
        catch (UnauthorizedAccessException)
        {
            return false;
        }
    }
}
